Readme.md
# koa-favicon [![Build Status](https://travis-ci.org/koajs/favicon.svg)](https://travis-ci.org/koajs/favicon)

 Koa middleware for serving a favicon. Based on [serve-favicon](https://github.com/expressjs/serve-favicon).

## Installation

```js
$ npm install koa-favicon
```

## Example

```js
const Koa = require('koa');
const favicon = require('koa-favicon');
const app = new Koa();

app.use(favicon(__dirname + '/public/favicon.ico'));
```

## API

### favicon(path, [options])

Returns a middleware serving the favicon found on the given `path`.

#### options

- `maxAge` cache-control max-age directive in ms, defaulting to 1 day.
- `mime` specify the mime-type of the favicon, defaults to "image/x-icon"

## License

  MIT


History.md
Unreleased
=================

 * Remove unused "mz" dependency

2.1.0 / 2020-02-27
==================

 * Support and test for specifying mime type of icon (#30)
 * bump deps

2.0.1 / 2018-03-18
==================

 * bump deps

1.2.0 / 2014-09-23
==================

 * use mz instead of co-fs

1.1.0 / 2014-05-05
==================

 * add favicon serving

1.0.1 / 2013-12-21
==================

 * update to new koa middleware signature


test/index.js
'use strict';

const request = require('supertest');
const join = require('path').join;
const assert = require('assert');
const favicon = require('..');
const Koa = require('koa');
const fs = require('fs');

describe('favicon()', function(){
  const path = join(__dirname, 'fixtures', 'favicon.ico');

  it('should only respond on /favicon.ico', done => {
    const app = new Koa();

    app.use(favicon(path));

    app.use((ctx, next) => {
      assert(!ctx.body);
      assert(!ctx.get('Content-Type'));
      ctx.body = 'hello';
    });

    request(app.listen())
    .get('/')
    .expect('hello', done);
  });

  it('should only respond on /favicon.ico if `path` is missing', done => {
    const app = new Koa();

    app.use(favicon());

    app.use((ctx, next) => {
      assert(!ctx.body);
      assert(!ctx.get('Content-Type'));

      ctx.body = 'hello';
    });

    request(app.listen())
    .get('/')
    .expect('hello', done);
  });

  it('should 404 if `path` is missing', done => {
    const app = new Koa();
    app.use(favicon());
    request(app.listen())
    .post('/favicon.ico')
    .expect(404, done);
  });

  it('should accept OPTIONS requests', done => {
    const app = new Koa();
    app.use(favicon(path));

    request(app.listen())
    .options('/favicon.ico')
    .expect('Allow', 'GET, HEAD, OPTIONS')
    .expect(200, done);
  });

  it('should not accept POST requests', done => {
    const app = new Koa();
    app.use(favicon(path));

    request(app.listen())
    .post('/favicon.ico')
    .expect('Allow', 'GET, HEAD, OPTIONS')
    .expect(405, done);
  });

  it('should send the favicon', done => {
    const body = fs.readFileSync(path);

    const app = new Koa();
    app.use(favicon(path));

    request(app.listen())
    .get('/favicon.ico')
    .expect('Content-Type', 'image/x-icon')
    .expect(200, body, done);
  });

  it('should send the favicon with a different mime type', done => {
    const body = fs.readFileSync(path);

    const app = new Koa();
    app.use(favicon(path, {mime: 'image/png'}));

    request(app.listen())
    .get('/favicon.ico')
    .expect('Content-Type', 'image/png')
    .expect(200, body, done);
  });

  it('should set cache-control headers', done => {
    const app = new Koa();
    app.use(favicon(path));
    request(app.listen())
    .get('/favicon.ico')
    .expect('Cache-Control', 'public, max-age=86400')
    .expect(200, done);
  });

  describe('options.maxAge', function(){
    it('should set max-age', done => {
      const app = new Koa();
      app.use(favicon(path, { maxAge: 5000 }));
      request(app.listen())
      .get('/favicon.ico')
      .expect('Cache-Control', 'public, max-age=5')
      .expect(200, done);
    });

    it('should accept 0', done => {
      const app = new Koa();
      app.use(favicon(path, { maxAge: 0 }));
      request(app.listen())
      .get('/favicon.ico')
      .expect('Cache-Control', 'public, max-age=0')
      .expect(200, done);
    });

    it('should be valid delta-seconds', done => {
      const app = new Koa();
      app.use(favicon(path, { maxAge: 1234 }));
      request(app.listen())
      .get('/favicon.ico')
      .expect('Cache-Control', 'public, max-age=1')
      .expect(200, done);
    });

    it('should floor at 0', done => {
      const app = new Koa();
      app.use(favicon(path, { maxAge: -4000 }));
      request(app.listen())
      .get('/favicon.ico')
      .expect('Cache-Control', 'public, max-age=0')
      .expect(200, done);
    });

    it('should ceil at 31556926', done => {
      const app = new Koa();
      app.use(favicon(path, { maxAge: 900000000000 }));
      request(app.listen())
      .get('/favicon.ico')
      .expect('Cache-Control', 'public, max-age=31556926')
      .expect(200, done);
    });

    it('should accept Infinity', done => {
      const app = new Koa();
      app.use(favicon(path, { maxAge: Infinity }));
      request(app.listen())
      .get('/favicon.ico')
      .expect('Cache-Control', 'public, max-age=31556926')
      .expect(200, done);
    });
  })
})


index.js
'use strict';

/**
 * Module dependencies.
 */

const resolve = require('path').resolve;
const fs = require('fs');

/**
 * Serve favicon.ico
 *
 * @param {String} path
 * @param {Object} [options]
 * @param {Number} [options.maxAge=null]
 * @param {String} [options.mime="image/x-icon"] MIME type of the file at path
 * @return {Function}
 * @api public
 */

module.exports = function (path, options){
  if (!path) {
    return (ctx, next) => {
      if ('/favicon.ico' != ctx.path) {
        return next();
      }
    };
  }

  path = resolve(path);
  options = options || {};

  let icon;
  const maxAge = options.maxAge == null
    ? 86400000
    : Math.min(Math.max(0, options.maxAge), 31556926000);
  const cacheControl = `public, max-age=${maxAge / 1000 | 0}`;
  const mime = options.mime || 'image/x-icon';

  return (ctx, next) => {
    if ('/favicon.ico' != ctx.path) {
      return next();
    }

    if ('GET' !== ctx.method && 'HEAD' !== ctx.method) {
      ctx.status = 'OPTIONS' == ctx.method ? 200 : 405;
      ctx.set('Allow', 'GET, HEAD, OPTIONS');
    } else {
      // lazily read the icon
      if (!icon) icon = fs.readFileSync(path);
      ctx.set('Cache-Control', cacheControl);
      ctx.type = mime;
      ctx.body = icon;
    }
  };
};
